freqtrade/exchange/kraken.py

```diff
@@ -0,0 +1,171 @@
+""" Kraken exchange subclass """
+import logging
+from random import randint
+from typing import Dict
+
+import arrow
+import ccxt
+
+from freqtrade import OperationalException, DependencyException, TemporaryError
+from freqtrade.exchange import Exchange
+
+logger = logging.getLogger(__name__)
+
+
+class Kraken(Exchange):
+
+    _params: Dict = {"trading_agreement": "agree"}
+
+    def __init__(self, config: dict) -> None:
+        super().__init__(config)
+
+    def buy(self, pair: str, ordertype: str, amount: float,
+            rate: float, time_in_force) -> Dict:
+        if self._conf['dry_run']:
```

Comment:
I would try to implement this in a different way, since 90% of this function is copy/pasted from the parent:

* extend the relevant functions in the parent class with an optional "params" parameter (defaulting to an empty dict?), which is passed to the underlying ccxt functions
* in the child-class (this one), add the necessary parameters and call `super().buy()`

This should reduce the amount of duplicate code - which is never a good thing as it means fixing a simple thing are changes in 20 classes (and it's difficult to keep track if the 21st has been forgotten).


a sample may look as follows (assuming the parent has been adjusted):
```python
def buy(self, pair: str, ordertype: str, amount: float,
            rate: float, time_in_force) -> Dict:

    params = self._params.copy()
    super().buy(pair, ordertype, amount, rate, time_in_force, params)
```

The parent class would also take care of all the exception handling and such...

Reply:
While making the necessary changes, I was wondering if an additional argument should be added to `buy()` and the other methods or if I should add a `_params` attribute to the base class which defaults to an empty dict and is used in the `create_order` calls. Then in the subclass only the attribute has to be changed and `buy` and `sell` could stay the same. It'd still be possible to adapt the `stoploss_limit` method. What do you think?

Reply:
it's definitely a possibility, yes
but does this then even need a subclass? 

adding an exchange configuration option (maybe `order_base_params` or `base_order_params`) and using that (a copy of it) in all order-creation/cancellation created methods?
it could be more flexible (not limited to kraken, should ExchangeXYZ require a similar parameter) - without requiring one class per exchange...
but on the other hand limiting if the parameters are different between buy and sell ...

Reply:
An additional note - if kraken supports stoploss orders (on exchange - as we call them) - then the subclass would make absolute sense, as i'm quite certain that it's not identical to binance...

Reply:
> but does this then even need a subclass?

I was thinking the same, but for the same reasons you're mentioning (stoploss etc.) I think it would make sense to have subclasses. I think there are various Binance specific things in the code already, for example `time_in_force`, which could also be moved to a subclass.

This is always the issue when using `ccxt`... It's a great library, but it's not standardised enough (yet), so you either have to go for only a few exchanges or make exchange specific adjustments..